pkg/cmd/pulumi/state_edit.go

```diff
@@ -0,0 +1,269 @@
+		Colorizer: cmdutil.GetGlobalColorization(),
+	}
+	cmd := &cobra.Command{
+		Use: "edit",
+		// TODO(dixler) Add YAML + comments with sorted keys before exposing this to users.
+		Hidden: !hasExperimentalCommands(),
+		Short:  "Edit the current stack's state in your EDITOR",
+		Long: `[EXPERIMENTAL] Edit the current stack's state in your EDITOR
+
+This command can be used to surgically edit a stack's state in the editor
+specified by the EDITOR environment variable and will provide the user with
+a preview showing a diff of the altered state.`,
+		Args: cmdutil.NoArgs,
+		Run: cmdutil.RunResultFunc(func(cmd *cobra.Command, args []string) result.Result {
+			if !cmdutil.Interactive() {
+				return result.Error("pulumi state edit must be run in interactive mode")
+			}
+			s, err := requireStack(commandContext(), stackName, stackLoadOnly, display.Options{
+				IsInteractive: true,
+			})
+			if err != nil {
+				return result.FromError(err)
+			}
+			if err := stateEdit.Run(s); err != nil {
+				return result.FromError(err)
+			}
+			return nil
+		}),
+	}
+	cmd.PersistentFlags().StringVar(
+		&stackName, "stack", "",
+		"Remove the stack and its config file after all resources in the stack have been deleted")
+	return cmd
+}
+
+type stateEditCmd struct {
+	Stdin     io.Reader
+	Stdout    io.Writer
+	Colorizer colors.Colorization
+	Ctx       context.Context
+}
+
+type snapshotBuffer struct {
+	Name     func() string
+	Snapshot func() (*deploy.Snapshot, error)
+	Reset    func() error
+	Cleanup  func()
+}
+
+func newSnapshotBuffer(fileExt string, sf snapshotEncoder, snap *deploy.Snapshot) (*snapshotBuffer, error) {
+	tempFile, err := os.CreateTemp("", "pulumi-state-edit-*"+fileExt)
+	if err != nil {
+		return nil, err
+	}
+	tempFile.Close()
+
+	originalText, err := sf.SnapshotToText(snap)
+	if err != nil {
+		// Warn that the snapshot is already hosed.
+		cmdutil.Diag().Errorf(diag.RawMessage("", fmt.Sprintf("initial state unable to be serialized: %v", err)))
+	}
+	t := &snapshotBuffer{
+		Name: func() string { return tempFile.Name() },
+		Snapshot: func() (*deploy.Snapshot, error) {
+			b, err := os.ReadFile(tempFile.Name())
+			if err != nil {
+				return nil, err
+			}
+			return sf.TextToSnapshot(snapshotText(b))
+		},
+		Reset: func() error {
+			return os.WriteFile(tempFile.Name(), originalText, 0o600)
+		},
+		Cleanup: func() {
+			os.Remove(tempFile.Name())
+		},
+	}
+	if err := t.Reset(); err != nil {
+		t.Cleanup()
+		return nil, err
+	}
+	return t, nil
+}
+
+func (cmd *stateEditCmd) Run(s backend.Stack) error {
+	if cmd.Stdin == nil {
+		cmd.Stdin = os.Stdin
+	}
+	if cmd.Stdout == nil {
+		cmd.Stdout = os.Stdout
+	}
+
+	snap, err := s.Snapshot(cmd.Ctx, stack.DefaultSecretsProvider)
+	if err != nil {
+		return err
+	} else if snap == nil {
+		return fmt.Errorf("old snapshot expected to be non-nil")
+	}
+
+	sf := &jsonSnapshotEncoder{
+		ctx: cmd.Ctx,
+	}
+	fileExt := ".json"
+	f, err := newSnapshotBuffer(fileExt, sf, snap)
```

Comment:
minor: can reduce number of variables in scope here; fileExt less useful now

```suggestion
	f, err := newSnapshotBuffer(".json", sf, snap)
```